core/client/fs/src/main/java/alluxio/client/metrics/MetricsHeartbeatContext.java

```diff
@@ -0,0 +1,241 @@
+    mConf = ctx.getConf();
+    // setup metrics master client sync
+    mMetricsMasterClient = new MetricsMasterClient(MasterClientContext
+        .newBuilder(ctx)
+        .setMasterInquireClient(inquireClient)
+        .build());
+    mClientMasterSync = new ClientMasterSync(mMetricsMasterClient, mConf);
+  }
+
+  private synchronized void addContext(String appId) {
+    mClientMasterSync.addAppId(appId);
+    if (mCtxCount.getAndIncrement() == 0) {
+      mMetricsMasterHeartbeatTask =
+          sExecutorService.scheduleWithFixedDelay(mClientMasterSync::heartbeat, 0,
+              mConf.getMs(PropertyKey.USER_METRICS_HEARTBEAT_INTERVAL_MS), TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private synchronized void heartbeat() {
+    mClientMasterSync.heartbeat();
+  }
+
+  /**
+   * Remove an application from this metrics heartbeat.
+   *
+   * A user who calls this method should assume the reference to this context is invalid
+   * afterwards. It will automatically close and remove itself from all tracking if the number
+   * of open contexts for this heartbeat reaches 0. Never attempt to add another context with
+   * the same reference after removing.
+   *
+   * @param appId the applicationId to remove
+   */
+  private synchronized void removeContext(String appId) {
+    mClientMasterSync.removeAppId(appId);
+    if (mCtxCount.decrementAndGet() <= 0) {
+      close();
+    }
+  }
+
+  /**
+   * When closed, this method will remove its task from the scheduled executor.
+   *
+   * It will also remove itself from being tracked in the MASTER_METRICS_HEARTBEAT. It should
+   * never need to be called manually.
+   */
+  private synchronized void close() {
+    mMetricsMasterClient.close();
+    if (mMetricsMasterHeartbeatTask != null) {
+      mMetricsMasterHeartbeatTask.cancel(false);
+    }
+    MASTER_METRICS_HEARTBEAT.remove(mMasterAddr);
+  }
+
+  /**
+   * Sets up a new metrics heartbeat with the given client information.
+   *
+   * This will instantiate a new executor service if it is the first heartbeat to be added,
+   * otherwise the application Id is simply included in an already existing heartbeat. This helps
+   * to consolidate RPCs and utilize less resources on the client.
+   *
+   * @param ctx The application's client context
+   * @param inquireClient the master inquire client used to connect to the master
+   * @param appId the application id that is associated with the metrics
+   */
+  public static synchronized void addHeartbeat(ClientContext ctx,
+      MasterInquireClient inquireClient, String appId) {
+    Preconditions.checkNotNull(ctx);
+    Preconditions.checkNotNull(inquireClient);
+    Preconditions.checkNotNull(appId);
+    if (sExecutorService == null) {
+      sExecutorService = Executors.newSingleThreadScheduledExecutor(
+          ThreadFactoryUtils.build("metrics-master-heartbeat-%d", true));
+
+      // register the shutdown hook if it hasn't been set up already
+      if (!sAddedShudownHook) {
```

Comment:
Should this be in its own condition instead of in the `sExecutorService == null` branch?

Reply:
It should only ever execute for one time in the life of the JVM, being within the `sExecutorService == null` branches just saves a condition check every time we call `addHeartbeat`. We're guaranteed on the first call to `addHeartbeat` that `sExecutorService` is `null`, so that check will always happen at least once.

Reply:
Right, I don't think there's anything wrong with the logic. From a code legibility perspective it is not obvious why these two conditions are coupled, and if its just a condition check we are saving, it seems more worthwhile to make the logic decoupled.